src/AspNet.Security.OAuth.Introspection/Events/ValidateAudienceContext.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace AspNet.Security.OAuth.Introspection {
+    public class ValidateAudienceContext : BaseIntrospectionContext {
+        public ValidateAudienceContext(
+            HttpContext context, 
+            OAuthIntrospectionOptions options,
+            JToken payload) 
+            : base(context, options) {
+            Payload = payload;
+            var audiences = payload[OAuthIntrospectionConstants.Claims.Audience];
+            if(audiences != null) {
+                if(audiences.Type == JTokenType.String) {
+                    Audiences = new[] { payload.Value<string>(OAuthIntrospectionConstants.Claims.Audience) };
+                }
+                else if (audiences.Type == JTokenType.Array) {
+                    Audiences = payload.Value<JArray>(OAuthIntrospectionConstants.Claims.Audience)
+                                           .Select(audience => audience.Value<string>());
+                }
+            }
+            Validate();
+        }
+
+        public IEnumerable<string> Audiences { get; set; }
+
+        public JToken Payload { get; }
+
+        public bool IsValid { get; set; }
+
+        public bool Validate() {
+            return IsValid = Validate(Audiences);
+        }
+
+        public bool Validate(IEnumerable<string> audiences)
+        {
+            IsValid = (audiences == null && Options.Audiences.Count > 0) || Options.Audiences.Count == 0 ||
```

Comment:
Actually, the more I think about it, the more I wonder if we really need such a specialized event. One is totally free to leave `Options.Audiences` empty to disable the audiences validation and to use the `TokenValidated` event to add custom validation logic, like a custom audience check.